ml-algorithms/src/main/java/org/opensearch/ml/engine/ingest/AbstractIngestion.java

```diff
@@ -0,0 +1,215 @@
+import org.opensearch.ml.common.utils.StringUtils;
+
+import com.jayway.jsonpath.JsonPath;
+
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+public class AbstractIngestion implements Ingestable {
+    public static final String OUTPUT = "output";
+    public static final String INPUT = "input";
+    public static final String OUTPUTIELDS = "output_names";
+    public static final String INPUTFIELDS = "input_names";
+    public static final String INGESTFIELDS = "ingest_fields";
+    public static final String IDFIELD = "id_field";
+
+    private final Client client;
+
+    public AbstractIngestion(Client client) {
+        this.client = client;
+    }
+
+    protected ActionListener<BulkResponse> getBulkResponseListener(
+        AtomicInteger successfulBatches,
+        AtomicInteger failedBatches,
+        CompletableFuture<Void> future
+    ) {
+        return ActionListener.wrap(bulkResponse -> {
+            if (bulkResponse.hasFailures()) {
+                failedBatches.incrementAndGet();
+                future.completeExceptionally(new RuntimeException(bulkResponse.buildFailureMessage()));  // Mark the future as completed
+                // with an exception
+                return;
+            }
+            log.debug("Batch Ingestion successfully");
+            successfulBatches.incrementAndGet();
+            future.complete(null); // Mark the future as completed successfully
+        }, e -> {
+            log.error("Failed to Batch Ingestion", e);
+            failedBatches.incrementAndGet();
+            future.completeExceptionally(e);  // Mark the future as completed with an exception
+        });
+    }
+
+    protected double calculateSuccessRate(List<Double> successRates) {
+        return successRates
+            .stream()
+            .min(Double::compare)
+            .orElseThrow(
+                () -> new OpenSearchStatusException(
+                    "Failed to batch ingest data as not success rate is returned",
+                    RestStatus.INTERNAL_SERVER_ERROR
+                )
+            );
+    }
+
+    /**
+     * Filters fields in the map where the value contains the specified source index as a prefix.
+     *
+     * @param mlBatchIngestionInput The MLBatchIngestionInput.
+     * @param index    The source index to filter by.
+     * @return A new map with only the entries that match the specified source index.
+     */
+    protected Map<String, Object> filterFieldMapping(MLBatchIngestionInput mlBatchIngestionInput, int index) {
+        Map<String, Object> fieldMap = mlBatchIngestionInput.getFieldMapping();
+        int indexInFieldMap = index + 1;
+        String prefix = "source[" + indexInFieldMap + "]";
+
+        Map<String, Object> filteredFieldMap = fieldMap.entrySet().stream().filter(entry -> {
+            Object value = entry.getValue();
+            if (value instanceof String) {
+                return ((String) value).contains(prefix);
+            } else if (value instanceof List) {
```

Comment:
Do we only support two types: String and List ? What if it's other type?

Reply:
It's scoped to be either a String or a List in the field mapping. key: value(String) or key: []. This should cover all cases.